### ISSUE TITLE
ming024 기반 FastSpeech2 구현 관련하여 질문 드립니다.

### ISSUE THREAD

**Issue by LEEYOONHYUNG** [HGU-DLLAB/Korean-FastSpeech2-Pytorch#16]
안녕하세요, 먼저 좋은 FastSpeech2 구현체를 공개해주시고 연구에 많은 도움을 주셔서 감사하다는 말씀 드립니다.

저도 현재 ming024 repo를 기반으로 FastSpeech2를 구현해보고 있는데 **생성 샘플에 잡음이 생기는 것**과, **pitch/energy를 frame-level로 사용하는 경우 음질이 크게 떨어진다는 점**. 이 두 가지 문제를 해결하고자 제 개인적인 노력을 추가로 진행하고 있습니다.

그러던 와중 본 repo를 보게 되었는데 본 repo에서 제공하는 sample은 퀄리티가 더욱 개선된 것으로 보이고, README.md에 나와있는 이미지는 pitch/energy를 frame-level로 사용하고 있는 것 같아 몇가지 질문을 드려보고 싶어서 이슈 남기게 됐습니다.

1. ming024와 본 repo를 비교해보면 데이터 전처리와 variance 정보를 사용하는 방법 등 몇몇 부분에서 차이가 나는 것 같은데요. 혹시 잡음 제거 및 음질 개선을 이루는데 유의미한 도움을 준 구현 요소가 있다면 조언 주실 수 있을까요? (e.g. 데이터 전처리, CNN embedding producer 사용, HIFI-GAN이 아닌 VocGAN 사용 등)

(README.md에 **[완료] pitch, energy loss가 total loss의 대부분을 차지하여 개선 중에 있음. [완료] 생성된 음성에서의 기계음 문제** 이 두 가지 해결 사항이 관련이 있을 것도 같은데 혹시 맞을까요?)

2. 현재 코드는 pitch/energy를 phoneme-level로 averaging하여 사용하는 것으로 보이는데 README.md에 나온 이미지는 그런 것 같지 않아서 질문 드립니다. 혹시 frame-level로 구현하여도 좋은 성능을 확인하셨을까요?

3. FastSpeech2의 경우 overfitting이 빠르게 발생하며, 그럼에도 불구하고 학습을 더 시킬 수록 음질이 개선되는 것으로 보이는데요. 혹시 이에 대한 insight가 있으실까요?

질문이 많아 죄송하지만, 조언 주신다면 정말 큰 도움 될 것 같습니다.
다시 한번 좋은 구현 공개해주셔서 감사하다는 말씀 드립니다.

**Comment by Jackson-Kang** [comment]
안녕하세요, @LEEYOONHYUNG 님.
저술하신 BVAE-TTS 흥미롭게 잘 읽었습니다.

말씀하신 문제를 해결하는데 제가 도움이 될 지 모르겠지만, 
제가 경험한 것과 아는 바에 근거해 최대한 답변드립니다.
(답변이 길어 질문 별로 보실 수 있도록 답변을 나누겠습니다.)


Q-1) ming024와 본 repo를 비교해보면 데이터 전처리와 variance 정보를 사용하는 방법 등 몇몇 부분에서 차이가 나는 것 같은데요. 혹시 잡음 제거 및 음질 개선을 이루는데 유의미한 도움을 준 구현 요소가 있다면 조언 주실 수 있을까요? (e.g. 데이터 전처리, CNN embedding producer 사용, HIFI-GAN이 아닌 VocGAN 사용 등)

A-1) 아래 사항들이 도움이 되었습니다. (음질 개선 순서대로 나열했습니다.)

1. 한국어를 alphabet 기반의 발음기호로 변경하지 않고, 한국어 초-중-종성으로 구성된 phoneme으로 Montreal Forced Aligner(MFA)를 학습 후 alignment 추출
      alphabet으로 표현된 한국어 발음기호(Phoneme)을 input으로 MFA를 학습시켰더니, duration이 외국인의 그것과 비슷하게 나왔습니다. (실제 학습 결과도, 외국인이 억지로 한국어를 발음하는 것 같은 음성이 합성되는 현상이 있었습니다.) 또한, FastSpeech2의 구조를 잘 분석해보면, 정확한 duration을 추출하는 것이 굉장히 중요합니다. 왜냐하면, FastSpeech2(FS2)는 합성시 예측된 duration만큼 길이가 길어진 frame-level prosody를 예측하여 반영하는데, duration이 잘못 예측되면 GT pitch/energy 의 격차가 더 커질 수 밖에 없습니다. 한국어 초-중-종성으로 구성된 phoneme으로 MFA를 학습 후 alignment 추출하여 학습시켰더니 문제가 해결되었고, pitch/energy의 예측 능력이 향상되는 동시에, 노이즈가 약간 감소하는 것을 경험하였습니다.

2. Neural vocoder(VocGAN)을 오랜 시간동안 학습
      경험적으로, TTS model(text2mel, neural vocoder 둘 다) 학습 시, loss의 minima를 찾았음에도 오래 학습시킬수록 (overfitting이 일어나도) 음질 등 개선이 되는 현상이 대부분 나타납니다. (아마 이미 경험하셔서 저보다 잘 아실거라 생각합니다.) 오래 학습시켜 음질의 개선을 이루었습니다.
      
3. pitch, energy를 normalize [-1, 1]의 범위를 갖게 함
      FastSpeech2는 GT pitch/energy를 이용해 음성 생성 시 pitch/energy를 disentangle시켜 controllability를 높이는 원리로 구현되어 있습니다. 그러므로, 당연히, pitch/energy predictor가 GT 수준의 pitch/energy를 예측하지 못할수록, decoder가 생성해내는 음성의 음질은 떨어질 수 밖에 없습니다. 결국 음질을 높이기 위해서는 최대한 GT 수준으로 pitch/energy sequence를 예측하는 것이 핵심이라고 생각했습니다. 그러므로, pitch/energy에 대해 normalization을 수행하여 pitch/energy predictor의 예측 능력을 향상시켰습니다. 

4. CNN embedding producer 사용
       원래 ming024님의 구현에는, training dataset의 pitch & energy를 quantization 후 embedding table을 사용하는 형태로 구현되어 있습니다. 각 구간 별로 독립적인 파라미터를 사용하면, min-max quantization의 특성 상, 특정 영역대의 pitch/energy값에 대응되는 embedding table이 학습되지 않아 (pitch/energy의 distribution에서 평균으로부터 멀 수록 outlier가 생길 확률이 높아지기 때문입니다.) 사실상 noise에 가까운 embedding vector로 남아있게 됩니다. (경험상, 합성 시 noise에 가깝게 모델링 된 구간에 해당하는 embedding을 반영하였더니 음질 저하가 심했습니다.) CNN embedding producer를 사용하면, 파라미터를 여러 prosody value에 대해 공유할 수 있고 local한 정보를 더 잘 반영할 수 있기 때문에, "조금 더" robust한 음성을 생성할 수 있을거라 판단했습니다. 실험 결과, 약간의 음질 개선 효과가 있었습니다.

**Comment by Jackson-Kang** [comment]
Q-2) 현재 코드는 pitch/energy를 phoneme-level로 averaging하여 사용하는 것으로 보이는데 README.md에 나온 이미지는 그런 것 같지 않아서 질문 드립니다. 혹시 frame-level로 구현하여도 좋은 성능을 확인하셨을까요?

A-2) README.md에 있는 figure들은 phoneme-level prosody를 반영하기 이전의 실험 결과들입니다. 처음 구현했던 것이 frame-level로 구현하였는데, 성능에 부족함을 느껴 phoneme-level prosody를 반영하도록 변경하였습니다.


Q-3) FastSpeech2의 경우 overfitting이 빠르게 발생하며, 그럼에도 불구하고 학습을 더 시킬 수록 음질이 개선되는 것으로 보이는데요. 혹시 이에 대한 insight가 있으실까요?

A-3) 개인적인 경험에 근거해 얘기드립니다. 
      경험상, 제가 학습해본 TTS 모델 중에서도 overfitting이 심한 모델들이 있었는데, 성능이 더 좋은 경우도 많았습니다. 자세한 이유는 저도 정확히 모르겠으나, 이 현상은 FastSpeech2에 국한되는 사항은 아닌 듯 합니다. 몇몇 논문들이 "딥러닝 모델은 loss의 minima = 최고 성능이 아니다. 더 오래 학습시키면 좋은 성능을 얻을 수 있다."는 주장을 하고 있는 것으로 알고 있습니다. (오래 전에 읽어 기억이 나지 않습니다.. 지금은 시간이 없어 나중에 찾게 되면 알려드릴게요!) 


부족하지만, 제 답변이 연구하시는데 도움이 되었길 바랍니다.
감사드립니다.

**Comment by LEEYOONHYUNG** [comment]
안녕하세요, @Jackson-Kang 님. 음질 개선 순서까지 알려주셔서 정말 큰 도움 됐습니다. 감사합니다.

실제로 duration의 중요성에 집중해서 모델을 살펴보던 중 MFA도 100% 정확한 duration을 뽑아주는 것은 아니라는 것을 확인하였고 해당 샘플들을 학습에서 제외한 후 학습을 돌리고 있는데 val_loss가 더 작은 지점에서 수렴하는 것으로 보이고 있네요.

이 외에도 overfitting과 관련한 좋은 경험 등 다양한 점들을 공유해주셔서 정말 감사합니다!!! :D

**Comment by Jackson-Kang** [comment]
안녕하세요, @LEEYOONHYUNG 님.
제 답변이 도움이 되셨다니 다행입니다.

다른 궁금한 사항이 있으면 언제든 문의 주시면 감사하겠습니다!
본 issue는 close하겠습니다.

감사드립니다.

**Comment by Moon-sung-woo** [comment]
안녕하세요, @LEEYOONHYUNG 님. 
위에 올려주신 글 보면 MFA가 100% 정확한 duration을 뽑아주지 못해 해당 샘플들을 학습에서 제외한 후 학습을 돌리셨다고 하신 것을 봤습니다.
MFA가 duration을 정확히 뽑아줬는지 아닌지를 확인하는 방법이 있는 것일까요? 아님 수작업으로 하신건가요??
어떻게 하셨는지 궁금하여 질문 남깁니다.
감사합니다.

**Comment by LEEYOONHYUNG** [comment]
안녕하세요 @Moon-sung-woo 님. 해당 부분은 제가 스스로 spectrogram과 phoneme duration을 함께 plot한 후 정말 이상하다 싶은 샘플들을 몇개 살펴봐서 자체적인 데이터 필터링 코드를 작성했던 걸로 기억합니다. (duration이 비정상적으로 긴 샘플들이 있었던 걸로 기억하네요.) 다만 구체적인 필터링 기준은 기억이 나질 않습니다.

**Comment by Moon-sung-woo** [comment]
안녕하세요 @LEEYOONHYUNG 님! 친절한 답변 너무 감사합니다. 이슈는 닫도록 하겠습니다.
감사합니다!

**Comment by Moon-sung-woo** [comment]
안녕하세요 @Jackson-Kang 님.
합성음의 품질을 올리는 방법 올려주셔서 감사합니다.
품질음을 올리는 방법중 궁금한 것이 있어 issue남기게 되었습니다.

'2. Neural vocoder(VocGAN)을 오랜 시간동안 학습' 하셔서 성능을 올리셨다 하셨는데
vocoder학습 방법에 대해 질문이 있습니다.

공개되어 있는 모델(예 hifi-GAN)의 universal한 pre-trained모델을 이용하여 fine-tuning을 할 예정입니다.
이때 단일화자의 고품질 합성음을 얻고싶다면 단일화자의 dataset만을 이용하여 fine-tuning 시키는 것이 좋을지
아님 한국인들의 다양한 dataset과 얻고자 하는 화자의 dataset을 합쳐 학습시키는 것이 좋은지 모르겠습니다.

바쁘실텐데 항상 자세하고 친절한 답변 달아주셔 감사합니다.
